packages/block-library/src/image/index.php

```diff
@@ -241,44 +261,65 @@ function block_core_image_render_lightbox( $block_content, $block ) {
-	// the zoom seamlessly on slow internet connections; the responsive
-	// image is a copy of the one in the body, which animates immediately
-	// as the lightbox is opened, while the enlarged one is a full-sized
-	// version that will likely still be loading as the animation begins.
-	$m = new WP_HTML_Tag_Processor( $block_content );
-	$m->next_tag( 'figure' );
-	$m->add_class( 'responsive-image' );
-	$m->next_tag( 'img' );
-	// We want to set the 'src' attribute to an empty string in the responsive image
-	// because otherwise, as of this writing, the wp_filter_content_tags() function in
-	// WordPress will automatically add a 'srcset' attribute to the image, which will at
-	// times cause the incorrectly sized image to be loaded in the lightbox on Firefox.
-	// Because of this, we bind the 'src' attribute explicitly the current src to reliably
-	// use the exact same image as in the content when the lightbox is first opened while
-	// we wait for the larger image to load.
-	$m->set_attribute( 'src', '' );
-	$m->set_attribute( 'data-wp-bind--src', 'context.core.image.imageCurrentSrc' );
-	$m->set_attribute( 'data-wp-style--object-fit', 'selectors.core.image.lightboxObjectFit' );
-	$initial_image_content = $m->get_updated_html();
-
-	$q = new WP_HTML_Tag_Processor( $block_content );
-	$q->next_tag( 'figure' );
-	$q->add_class( 'enlarged-image' );
-	$q->next_tag( 'img' );
-
-	// We set the 'src' attribute to an empty string to prevent the browser from loading the image
-	// on initial page load, then bind the attribute to a selector that returns the full-sized image src when
-	// the lightbox is opened. We could use 'loading=lazy' in combination with the 'hidden' attribute to
-	// accomplish the same behavior, but that approach breaks progressive loading of the image in Safari
-	// and Chrome (see https://github.com/WordPress/gutenberg/pull/52765#issuecomment-1674008151). Until that
-	// is resolved, manually setting the 'src' seems to be the best solution to load the large image on demand.
-	$q->set_attribute( 'src', '' );
-	$q->set_attribute( 'data-wp-bind--src', 'selectors.core.image.enlargedImgSrc' );
-	$q->set_attribute( 'data-wp-style--object-fit', 'selectors.core.image.lightboxObjectFit' );
-	$enlarged_image_content = $q->get_updated_html();
+	$end_of_img_tag = $image_at + strlen( $image_tag );
+	$body_content   = substr( $body_content, 0, $end_of_img_tag ) . $button . substr( $body_content, $end_of_img_tag );
+
+	/*
+	 * This code needs to generate a responsive image and an enlarged image
+	 * to animate zoom seamlessly on slow internet connections; the responsive
+	 * image is a copy of the one in the body, which animates immediately
+	 * as the lightbox is opened, while the enlarged one is a full-sized
+	 * version that will likely still be loading as the animation begins.
+	 *
+	 * In order to reuse the existing Tag Processor, changes made before
+	 * need to be undone before setting the new changes here.
+	 */
```

Comment:
I don't think it will make sense to include a `$reset_changes` mode to the tag processor because of the complexity and runtime costs that could introduce. I have explored a chunked mode that could be used here instead, but in this particular case I think most solutions will be a bit messy.

your concern is real though, and I share it. it's a bit nasty having to reset. I did run some initial profiling but it's hard to get a good handle on the runtime impact without a more typical representative page, something I didn't have time to do last week.